sklearn/impute/_iterative.py

```diff
@@ -646,19 +646,27 @@ def _initial_imputation(self, X, in_fit=False):
         else:
             X_filled = self.initial_imputer_.transform(X)
 
-        valid_mask = np.flatnonzero(
-            np.logical_not(np.isnan(self.initial_imputer_.statistics_))
-        )
-
         if not self.keep_empty_features:
+            valid_mask = np.flatnonzero(
+                np.logical_not(np.isnan(self.initial_imputer_.statistics_))
+            )
+
             # drop empty features
             Xt = X[:, valid_mask]
             mask_missing_values = mask_missing_values[:, valid_mask]
+
+            if in_fit:
+                is_empty_feature = np.all(mask_missing_values, axis=0)
+                mask_missing_values[:, is_empty_feature] = False
+                Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
+
         else:
             # mark empty features as not missing and keep the original
             # imputation
-            mask_missing_values[:, valid_mask] = True
+            is_empty_feature = np.all(mask_missing_values, axis=0)
+            mask_missing_values[:, is_empty_feature] = False
             Xt = X
+            Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]
```

Comment:
So I would expect the following:


```python

is_empty_feature = np.all(mask_missing_values, axis=0)

        if not self.keep_empty_features:
            # drop empty features
            Xt = X[:, ~is_empty_feature]
            mask_missing_values = mask_missing_values[:, ~is_empty_feature]

            if self.initial_imputer_.get_params()['strategy'] == 'constant':
                # The constant strategy has a specific behavior and preserve empty
                # features even with ``keep_empty_features=False``. We need to drop
                # the column for consistency.
                # TODO: remove this `if` branch once the following issue is addressed:
                # https://github.com/scikit-learn/scikit-learn/issues/29827
                X_filled = X_filled[:, ~is_empty_feature]
        else:
            # mark empty features as not missing and keep the original imputation
            mask_missing_values[:, is_empty_feature] = False
            Xt = X
            Xt[:, is_empty_feature] = X_filled[:, is_empty_feature]

```

Reply:
Tried with this change and got `IndexError: index 3 is out of bounds for axis 1 with size 3` on this test `test_iterative_imputer_skip_non_missing`.

Basically, it is when
- `X_train` has **no** empty features
- `X_test` has empty features
- `keep_empty_features=False`
- `skip_complete=True` or `False`

```python
X_train = np.array([[5, 2, 2, 1], [10, 1, 2, 7], [3, 1, 1, 1], [8, 4, 2, 2]])
X_test = np.array([[np.nan, 2, 4, 5], [np.nan, 4, 1, 2], [np.nan, 1, 10, 1]])
imputer = IterativeImputer(
    initial_strategy="mean", skip_complete=skip_complete, random_state=rng
)
X_test_est = imputer.fit(X_train).transform(X_test)
if skip_complete:
    # impute with the initial strategy: 'mean'
    assert_allclose(X_test_est[:, 0], np.mean(X_train[:, 0]))
else:
    assert_allclose(X_test_est[:, 0], [11, 7, 12], rtol=1e-4)
```

In `transform`, using only this 
```python
is_empty_feature = np.all(mask_missing_values, axis=0)
valid_mask = ~is_empty_feature
```
will remove the empty feature from `X_test` which I guess we don't expect since the `X_train` has no empty feature.

What do you think?

I can fix with these following changes:
```python
if in_fit:
    is_non_empty_feature = np.logical_not(is_empty_feature)
else:
    is_non_empty_feature = np.logical_not(np.isnan(self.initial_imputer_.statistics_))
``` 

But it still require `in_fit`. Let me know if you have another suggestion.

Reply:
> is_non_empty_feature = np.logical_not(np.isnan(self.initial_imputer_.statistics_))

I don't like so much this line because the `SimpleImputer` is not consistent. Maybe the cleaner way is to store the empty feature at `fit` and not recompute it otherwise:

```python
        if in_fit:
            self._is_empty_feature = np.all(mask_missing_values, axis=0)

        if not self.keep_empty_features:
            # drop empty features
            Xt = X[:, ~self._is_empty_feature]
            mask_missing_values = mask_missing_values[:, ~self._is_empty_feature]

            if self.initial_imputer_.get_params()['strategy'] == 'constant':
                # The constant strategy has a specific behavior and preserve empty
                # features even with ``keep_empty_features=False``. We need to drop
                # the column for consistency.
                # TODO: remove this `if` branch once the following issue is addressed:
                # https://github.com/scikit-learn/scikit-learn/issues/29827
                X_filled = X_filled[:, ~self._is_empty_feature]
        else:
            # mark empty features as not missing and keep the original imputation
            mask_missing_values[:, self._is_empty_feature] = False
            Xt = X
            Xt[:, self._is_empty_feature] = X_filled[:, self._is_empty_feature]
```

We don't really need to expose it so I think that we can keep it private by starting with an `_`.

Reply:
Addressed in https://github.com/scikit-learn/scikit-learn/pull/29779/commits/3d04df9fdb51b8c8f20e0cf1e45860470171c82b